### ISSUE TITLE
Since 24.7 Health reporting graphs are missing data for periods at a time.

### ISSUE THREAD

**Issue by pyrodex** [opnsense/core#7753]
**Describe the bug**

Since the upgrade from 24.1 to 24.7, and now also 24.7.1, I am seeing data going missing in the health reporting graphs regardless of which graph. Also when I hover on the graph for timestamps in the 1 minute granularity I can see it normally ~40 minutes when the graph stop showing data. For example:

STOPS at 06:37, RESUMES at 07:08
STOPS at 07:38, RESUMES at 08:08

The system is ONLINE and running the entire time when there is missing data. 

This was working without issues on 24.1 and all 24.1.X minor versions. 

**To Reproduce**

1. Upgraded from 24.1 to 24.7
2. Go to Reporting -> Health
3. Click on any graph, the system processor one is good to see this since it is always logging.
4. Observe missing data

**Expected behavior**

The graph shouldn't be missing data...

**Describe alternatives you considered**

I've performed a "Reset RRD Data", didn't fix it.
Deleted the system-processor database from the settings, didn't fix it.
**Screenshots**

This screenshot was just a normal shot:
![image](https://github.com/user-attachments/assets/360c27b9-346d-4294-9e8c-4e722d9ecf28)

This screenshot was AFTER I deleted the system-processor RRD database via the webUI and let it log for a few hours to demonstrate a FRESH database:
![image](https://github.com/user-attachments/assets/c333e2e9-f15e-4836-828e-9c0a05a60dae)

**Relevant log files**

Not sure where to pull this from but let me know.

**Additional context**

N/A

**Environment**

Software version used and hardware type if relevant, e.g.:

Running under Proxmox (latest version)
OPNsense 24.7.1 (amd64).
Intel® Xeon™ D-2123IT (8 cores, 8 threads)

**Comment by fichtner** [comment]
Could be a race condition that surfaced in FreeBSD 14. I know our code didn’t change. I’ll take a look next week.

Cheers,
Franco

**Comment by dstroot** [comment]
For what it's worth I get no data at all on 2.7.1. Fully running loaded system:

<img width="1103" alt="Screenshot 2024-08-09 at 8 52 59 PM" src="https://github.com/user-attachments/assets/e678ab54-b436-4ea4-ac18-23073785d727">

**Comment by AdSchellevis** [comment]
@dstroot can you check if rrd is running on your end? (`ps fax | grep rrd`). On my end it seems to have died somewhere in between, maybe because I fiddled a bit with connectors last night causing various up/down events on interfaces.

The code unfortunately is very old without much debug output, one day we should refactor this as well I'm afraid.

https://github.com/opnsense/core/blob/805489d1a0d9628c982a73974d3317151f17df31/src/etc/inc/rrd.inc#L45

**Comment by beneix** [comment]
I have the same issue on 24.7.1 and resetting RRD data has not fixed it.

**Comment by doktornotor** [comment]
> @dstroot can you check if rrd is running on your end? (`ps fax | grep rrd`). On my end it seems to have died somewhere in between, maybe because I fiddled a bit with connectors last night causing various up/down events on interfaces.

Yeah, not running here.

**Comment by dstroot** [comment]
@doktornotor just wanted to say hi - I remember reading lots of your posts on the pfSense forums. ;).

Here's output looking to see if RRD is running:

```
root@firewall:~ # ps fax | grep rrd
86584  0  S+     0:00.00 grep rrd
root@firewall:~ #
```

**Comment by doktornotor** [comment]
🍻

Yes, nothing running. Just re-saving the settings seems to kick it alive. It's been running for about half a day. Just now, I tried some fiddling, such as disabling and re-enabling IPS -  and sure enough, it's dead again.

**Comment by meyergru** [comment]
For me, rrd did not run. After disabling and re-enabling, the daemon started. It looks as if it randomly stops and sometimes starts again. I have no idea why it stops and I also do not see when/why it starts again. E.g., there is no break in the internet connection.

**Comment by doktornotor** [comment]
It appears to do zero logging, kinda hard to debug such things. Again, non started after reboot.

**Comment by AdSchellevis** [comment]
ancient code, I'll try to take a look tomorrow to see if we clean this up a bit and add some logging.

**Comment by Adrian-Grimm** [comment]
same here on a DEC740 since 24.7 but also with most recent 24.7.1-amd64 nothing changed.

<img width="1832" alt="image" src="https://github.com/user-attachments/assets/4746d18a-21ec-47cb-bd1c-2ab203236726">

**Comment by Adrian-Grimm** [comment]
Just to run this command a further time `/usr/sbin/daemon -f -p '/var/run/updaterrd.pid' '/var/db/rrd/updaterrd.sh'` fixed it. But only till a reboot...

**Comment by doktornotor** [comment]
You can just go to Reporting - Settings - Reporting Database Options and click Save. 

Looking at https://github.com/search?q=repo%3Aopnsense%2Fcore%20rrd_configure&type=code, I'd say the potential for some race condition is quite high here during boot,  Having this is `rc.newwanip`, `rc.newwanip6` and `rc.bootup` does not help. Looking at the boot sequence output:

```
Generating RRD graphs...done.
Starting power daemon...done.
>>> Invoking start script 'newwanip'
Reconfiguring IPv4 on ovpns1
Reconfiguring IPv4 on pppoe0
Reconfiguring IPv6 on pppoe0
```

I'd say it will get killed and restarted at least 3 times after it's been already initially set up (the script created) and started (Generating RRD graphs...). Hmmm, I don't get the logic.

**Comment by fichtner** [comment]
Always been this way. I added a bit of locking if I recall correctly but in FreeBSD 14.1 this doesn’t work reliably it seems.

**Comment by fichtner** [comment]
https://github.com/opnsense/core/commit/fdf46f317c34b70398a6dcb9487fe141bcfd55f9

**Comment by doktornotor** [comment]
> [fdf46f3](https://github.com/opnsense/core/commit/fdf46f317c34b70398a6dcb9487fe141bcfd55f9)

That looks like success, the number of processes has been reduced nicely - to zero. 🤣

**Comment by fichtner** [comment]
It is funny but also frustrating. 😉

**Comment by doktornotor** [comment]
Well, really not obvious what changed that's causing this. Here's a quick idea for a quick test - what's a good way to **not** run `rrd_configure()` in those `rc.newwanip{,6}` scripts when the system is booting? `exit_on_bootup()` ?

**Comment by AdSchellevis** [comment]
A longer term better option might be to just use cron to start a collector every minute, it's usually less fragile and the script has no real advantage of staying in memory anyway in its current state. Next we can start thinking about splitting this monster into manageable chunks.

**Comment by fichtner** [comment]
Biggest offender in this construct is the gateway quality graph which needs a special trigger, at least how it was designed. Not sure if really necessary with the right design.

**Comment by AdSchellevis** [comment]
I'll take a quick look if we can improve this a bit, also keeping the monitors in mind.

**Comment by AdSchellevis** [comment]
To try this, first run:
```
opnsense-patch d2bab9c
```

Then go to "Reporting: Settings" and save settings,

**Comment by pyrodex** [comment]
I applied the patch and will report back after a few hours.

**Comment by doktornotor** [comment]
Looks good here.

```
# grep rrd /var/cron/tabs/*
/var/cron/tabs/root:*   *       *       *       *       (test -e /var/db/rrd/updaterrd.sh && /usr/local/bin/flock -n -E 1 /var/db/rrd/updaterrd.sh /var/db/rrd/updaterrd.sh) > /dev/null
/var/cron/tabs/root:0   */24    *       *       *       (/usr/local/etc/rc.syshook.d/backup/20-rrd) > /dev/null
```

**Comment by pyrodex** [comment]
> To try this, first run:
> 
> ```
> 
> opnsense-patch 3e99b87
> 
> ```
> 
> 
> 
> Then go to "Reporting: Settings" and save settings, after which the stats should be collected via cron.

Do we just run this one or the other one as well? After running just this one I got no data nearly 30 minutes later still.

**Comment by AdSchellevis** [comment]
only this one and  don't forget to save....

**Comment by meyergru** [comment]
I applied the patch, saved reporting settings, but still seeing only:

```
# grep rrd /var/cron/tabs/*
/var/cron/tabs/root:0   */24    *       *       *       (/usr/local/etc/rc.syshook.d/backup/20-rrd) > /dev/null

```

**Comment by pyrodex** [comment]
> only this one and  don't forget to save.... 

I did and still nothing is logging or at least being reflected on the webUI.

**Comment by doktornotor** [comment]
Maybe your cron is dead as well? Works perfectly fine here. Run `ps aux | grep '[c]ron' && date && ls -la /var/db/rrd`

```
# ps aux | grep '[c]ron' && date && ls -la /var/db/rrd
root    43550   0.0  0.1   12824    2108  -  Is   13:45      0:00.22 /usr/sbin/cron -s
Sun Aug 11 14:53:30 CEST 2024
total 2817
drwxr-xr-x   2 nobody wheel     25 Aug 11 00:00 .
drwxr-xr-x  21 root   wheel     29 Aug 11 13:46 ..
-rw-r--r--   1 nobody wheel 147848 Aug 10 18:32 ModemAccess_GWv4-quality.rrd
-rw-r--r--   1 nobody wheel 147848 Aug 11 14:53 WAN_DHCP6-quality.rrd
-rw-r--r--   1 nobody wheel 147848 Aug 11 14:53 WAN_PPPOE-quality.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 11 14:53 ipsec-packets.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 11 14:53 ipsec-traffic.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 11 14:53 lan-packets.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 11 14:53 lan-traffic.rrd
-rw-r--r--   1 nobody wheel 882048 Aug 11 14:53 ntpd.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 11 14:53 opt1-packets.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 11 14:53 opt1-traffic.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 10 18:32 opt2-packets.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 10 18:32 opt2-traffic.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 11 14:53 ovpns1-packets.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 11 14:53 ovpns1-traffic.rrd
-rw-r--r--   1 nobody wheel  49720 Aug 11 14:53 ovpns1-vpnusers.rrd
-rw-r--r--   1 nobody wheel 195704 Aug 11 14:53 system-cputemp.rrd
-rw-r--r--   1 nobody wheel 588592 Aug 11 14:53 system-mbuf.rrd
-rw-r--r--   1 nobody wheel 735320 Aug 11 14:53 system-memory.rrd
-rw-r--r--   1 nobody wheel 245976 Aug 11 14:53 system-processor.rrd
-rw-r--r--   1 nobody wheel 245976 Aug 11 14:53 system-states.rrd
-rwxr-xr-x   1 root   wheel   7872 Aug 11 13:45 updaterrd.sh
-rw-r--r--   1 nobody wheel 393168 Aug 11 14:53 wan-packets.rrd
-rw-r--r--   1 nobody wheel 393168 Aug 11 14:53 wan-traffic.rrd
```

**Comment by meyergru** [comment]
No. It did not work for me as well. Restart the web gui before saving, after that, it works.

**Comment by doktornotor** [comment]
*shrug* No web gui restart needed here.

**Comment by pyrodex** [comment]
> No. It did not work for me as well. Restart the web gui before saving, after that, it works.

Same for me... I see data now coming in but let me watch to see if the gaps come back.

**Comment by AdSchellevis** [comment]
Only reason I can find in the code is using the wrong save button to be honest, assuming no crash report is being generated for broken code. When the rrd save is pressed, this is executed:

https://github.com/opnsense/core/blob/f3669a58d778afde7ed41b6822c9ebd5a3e34c5c/src/www/reporting_settings.php#L78-L81

**Comment by lostcontrol** [comment]
This one is painful. The RRD collection stops after a few hours and never comes back until I toggle "Enables the RRD graphing backend". The RRD update script just dies for some reason. Resetting the RRD data does not help.

**Comment by doktornotor** [comment]
> This one is painful. 

Perhaps see https://github.com/opnsense/core/issues/7753#issuecomment-2282723192 and test it.

**Comment by joeroback** [comment]
applied the patch to 24.7.1. went to settings, unchecked save. checked saved. got this
```
Fatal error: Uncaught TypeError: flock(): Argument #1 ($stream) must be of type resource, bool given in /usr/local/opnsense/mvc/app/library/OPNsense/Core/FileObject.php:50 Stack trace: #0 /usr/local/opnsense/mvc/app/library/OPNsense/Core/FileObject.php(50): flock(false, 2) #1 /usr/local/etc/inc/rrd.inc(113): OPNsense\Core\FileObject->__construct('/var/db/rrd/upd...', 'r+', 493, 2) #2 /usr/local/www/reporting_settings.php(79): rrd_configure() #3 {main} thrown in /usr/local/opnsense/mvc/app/library/OPNsense/Core/FileObject.php on line 50
```

**Comment by AdSchellevis** [comment]
@joeroback add this https://github.com/opnsense/core/commit/7a6f65a293f84b9a65b7392ad1d074efb5f7faab one on top and you should be fine. It looks like the script doesn't exist, might be caused by using a memdisk. We'll try to add this to 24.7.2

**Comment by joeroback** [comment]
> @joeroback add this [7a6f65a](https://github.com/opnsense/core/commit/7a6f65a293f84b9a65b7392ad1d074efb5f7faab) one on top and you should be fine. It looks like the script doesn't exist, might be caused by using a memdisk. We'll try to add this to 24.7.2

FWIW, this is a basic opnsense install on n305 with a single sata ssd with zfs stripe. nothing configured for ram disk under /var etc.

i applied the second patch and while the original error is gone, the reporting -> settings page hangs on save, as it appears the for/sleep loop is still in `/var/db/rrd/updaterrd.sh` at the bottom
```
 for UNUSED in $(seq 1 60); do sleep 1; done
```

**Comment by AdSchellevis** [comment]
then you didn't apply https://github.com/opnsense/core/commit/d2bab9c3c89c8862d37e43bc5bd394396eabb644 as it removes the loop...  keep in mind, running the same patch twice reverts it.

**Comment by joeroback** [comment]
correct.

i disabled reporting. verified the `update_rrd.sh` was not running via `ps ax | grep rrd`

```
opnsense-patch 3e99b87
opnsense-patch 7a6f65a
```

enabled reporting and pressed save. the resulting `/var/db/rrd/updaterrd.sh` had the update loop in it, the UI was spinning on the save. maybe some weird caching. not sure, but refreshing and toggling reporting on/off eventually regenerated the script correctly.

its been running good now for 1 hour. 


---

as far as https://github.com/opnsense/core/commit/d2bab9c3c89c8862d37e43bc5bd394396eabb644 goes, that removes the while loop, but not
```
for UNUSED in $(seq 1 60); do sleep 1; done
```
which was at the bottom of the file. not sure how or who puts that there.

**Comment by doktornotor** [comment]
> as far as [d2bab9c](https://github.com/opnsense/core/commit/d2bab9c3c89c8862d37e43bc5bd394396eabb644) goes, that removes the while loop, but not
> 
> ```
> for UNUSED in $(seq 1 60); do sleep 1; done
> ```
> 
> which was at the bottom of the file. not sure how or who puts that there.

Hmmmm? 
https://github.com/opnsense/core/commit/d2bab9c3c89c8862d37e43bc5bd394396eabb644#diff-b23957fd637204ed089219253801dfb58d5a50b63579107e46f89c44a378f2c8L545

**Comment by doktornotor** [comment]
While here, the indenting of the resulting _heredoc_ script (hopefully to go away soon) may be dangerous for people suffering from OCD. 😆  Had to triple check the `done` that's sticking there as a sore thumb has an opening `do` somewhere. Yeah, it does - with a space missing in front of it.

**Comment by Adrian-Grimm** [comment]
OPNsense 24.7.2 update fixed the issue in my case.

**Comment by fichtner** [comment]
@AdSchellevis is working on a new version but let me close this ticket since the scope is addressed with 24.7.2 indeed.
